### PR TITLE
Handle the parameter wrapping for SPMD

### PR DESCRIPTION
**Pull request by JackCaoG**
should fix https://github.com/pytorch/xla/issues/7161

### CODE REVIEW

**Comment by alanwaketan** [comment]
Let me know when it's ready for review.

**Comment by JackCaoG** [comment]
not sure why TPU CI is skipped, but I think this pr is ready for review.

File: torch_xla/csrc/helpers.cpp [hunk 1]
```diff
@@ -1007,6 +1018,26 @@ xla::StatusOr<xla::XlaComputation> XlaHelpers::WrapXlaComputation(
...
+      // we assume that parameter is ordered.
+      XLA_CHECK_EQ(index, param_shardings.size());
+      param_shardings.push_back(*xla::HloSharding::FromProto(instr.sharding()));
+      TF_VLOG(5) << "index = " << index
+                 << " shape = " << xla::Shape(instr.shape()).ToString()
+                 << " sharding = "
+                 << xla::HloSharding::FromProto(instr.sharding())->ToString();
+    }
+  }
+  return std::move(param_shardings);
```
Comment (alanwaketan) [inline torch_xla/csrc/helpers.cpp:1038]: I don't think we need to do std::move for return value.

File: test/spmd/test_spmd_parameter_wrapping.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    auto_wrap_policy = functools.partial(
+        transformer_auto_wrap_policy,
+        transformer_layer_cls={torch.nn.Linear},
+    )
+    linears = FSDPv2(linears, auto_wrap_policy=auto_wrap_policy)
+    input = torch.randn(100, 10, device=device)
+    output = linears(input)
+    torch_xla.sync()
+    xm.wait_device_ops()
+    self.assertEqual(output.shape, torch.Size([100, 40]))
```
Comment (alanwaketan) [inline test/spmd/test_spmd_parameter_wrapping.py:51]: Is this check sufficient? Without this change, what will be the output shape?

Comment (JackCaoG) [reply test/spmd/test_spmd_parameter_wrapping.py:51]: I was just lazy and want to make sure this does not crash haha. I can add a unit test for check the value.

Comment (alanwaketan) [reply test/spmd/test_spmd_parameter_wrapping.py:51]: I guess my point is I cannot link the test to the change as the change attach the sharding to inputs but then you are checking the output.

**Review by alanwaketan — APPROVED** [review]

File: test/run_tests.sh [hunk 3]
```diff
@@ -117,6 +117,11 @@ function run_pt_xla_debug {
   PT_XLA_DEBUG=1 PT_XLA_DEBUG_FILE="/tmp/pt_xla_debug.txt" run_test "$@"
 }
 
+function run_parameter_warpping {
+  echo "Running in parameter wrapping mode: $@"
+  XLA_PARAMETER_WRAPPING_THREADSHOLD=1 run_test "$@"
```
Comment (JackCaoG) [inline test/run_tests.sh:122]: now I think about it maybe it is better to set this env var as part of the test

**Comment by JackCaoG** [comment]
Let me fix the review comments in a follow up pr so we can land this change for @alanwaketan to try.

**Pull request merged**